util/exercise/src/structs.rs

```diff
@@ -0,0 +1,159 @@
+//! Rust struct for canonical test data.
+//!
+//! See https://github.com/exercism/problem-specifications/blob/master/canonical-schema.json
+//! for more details on the JSON schema, which makes it possible to implement
+//! `serde::Deserialize`.
+
+use crate::{self as exercise};
+use indexmap::IndexMap;
+use serde::{Deserialize, Serialize};
+
+macro_rules! pub_struct_serde_getters {
+    (
+        pub struct $name:ident {
+            $( $fname:ident : $ftype:ty ),* $(,)*
+        }
+    ) => {
+        #[derive(Serialize, Deserialize, Debug)]
+        pub struct $name {
+            $($fname : $ftype),*
+        }
+
+        impl $name {
+            $(
+            pub fn $fname(&self) -> &$ftype {
+                &self.$fname
+            }
+            )*
+        }
+    };
+}
+
+pub_struct_serde_getters! {
+    pub struct CanonicalData {
+        exercise: Exercise,
+        version: Version,
+        comments: Option<Comments>,
+        cases: TestGroup,
+    }
+}
+
+type Exercise = String;
+type Version = String;
+type Comments = Vec<String>;
+type TestGroup = Vec<LabeledTestItem>;
+
+#[derive(Serialize, Deserialize, Debug)]
+#[serde(untagged)]
+pub enum LabeledTestItem {
+    Single(LabeledTest),
+    Array(LabeledTestGroup),
+}
+
+pub_struct_serde_getters! {
+    pub struct LabeledTest {
+        description: Description,
+        optional: Option<Optional>,
+        comments: Option<Comments>,
+        property: Property,
+        input: Input,
+        expected: Expected,
+    }
+}
+
+impl LabeledTest {
+    pub fn generate_test_function(&self) -> String {
+        // Attributes.
+        let mut s = vec![String::from("#[test]"), String::from("#[ignore]")];
```

Comment:
No, you didn't miss it. While this obviously sometimes adds an `#[ignore]` unnecessarily, I feel it will probably be easier to handle as a final step: find the first test function and drop the `#[ignore]` (or even just deleting the first `#[ignore]` of the file, since I believe only the test functions have them).